### DISCUSSION TITLE
New cup idea

### DISCUSSION THREAD

**Discussion by PabloCorso** [elmadev/elmaonline-site#457] [General]
Hi, I would like to share a new cup idea and ask for help to know if it's feasible to perform it using EOL and elma site.  I've been putting down the idea and doing some tests in [this repository](https://github.com/PabloCorso/eol-decentralized-cup/blob/main/README.md).

### Cup idea

The main idea is to create a cup that is run by everyone at the same time, meaning that every player will be able to collaborate with levels at any point in time during the cup. These levels will have to pass a very basic validation before being added as a new ongoing event, like for example: is it a new level? can it be finished?

Levels are added as events right after being summited and accepted. All events would have the same deadline as the cup. If the cup lasts 1 month and ends on date D, all added events end on date D. There surely will be some limitations like "A player can submit up to 1 level per day". 

Given this rules the cup could end up with too many levels, and no one will be able to hoyla them all, thus being too hard to compete. A solution: when the cup ends, only small selection of levels will be taken into account for the results. This can be done by selecting in some way the most popular levels, the levels in which players put the most effort and hopefully the ones they had the most fun playing.

Right now the best idea that I got to achieve this selection is by ranking the levels and picking just the top X better ranked levels at the end of the cup. After some testing, a ranking that feels simple and good enough is the following:

> Level points are the sum of all finished times, except the finishes that  are above 2 times the best time in the level.

By removing the higher times we would be removing any lazy or spam finishes, for example an 1 hour finish in a 10 seconds level. Usually short levels will have more finishes than longer ones, but short levels will sum points at a much slower rate too, getting sort of a balanced ranking. The more you play and finish a level, the better it ranks. Here are some tests for this specific algorithm: [All finishes tests summary](https://github.com/PabloCorso/eol-decentralized-cup/blob/main/scripts/summaries/summary_all.md).

### Cup managing

There would be some minimum managing for the levels and levels ranking, this is how it would look like:

1. Cup starts without any events on it. The deadline is set to date D.
2. Player submits a level, sends it to me or anyone in charge. The level is validated manually.
3. The validated level is renamed to something like Des1C001.lev with a proper consistent title, and it is added as a new ongoing event with deadline D. Players and times are hidden at EOL.
4. The levels ranking is updated, since a new level was added. Also, this ranking should be updated and shared to everyone regularly, ideally every time a new finish is made.
5. The cup ends, only the top X levels are taken into account to calculate the results and winner.

#### Submitting levels
This could be a direct message in Discord, or mopolauta... Or using the File Upload at elma site? Could be coding something into the cup page too, to submit levels, if it makes sense.

#### Only top X levels matter for the results
Calculating the results from only top X ranked levels could be done manually, or could be coded into the site if it's easy and makes sense to do it. This would be having the ability to select the levels that should be taken into account for the final results when the cup ends. The points for the players are calculated as in other cups (e.g. 100 points for 1st position, then 85, 75, etc.) but only using those selected levels.

#### Ranking levels to find the top X
It could be also done "manually" by having some external site (or Discord channel) that I could update regularly, but would need a way to access all finishes in the cup levels which would be hidden. Also it might be already possible to do this if the cup didn't hide level times at all, but I don't know if it would be nice to have a cup with times shown.

--------
What do you think about this idea? Could this have any success? Any opinion and also technical help about this last points is very welcome :>

**Comment by j-masci** [comment]
Ok I did read but was a while ago. Can you explain how the current cup system works in regards to number of skips and the admin assigning points manually (is that possible for example). If we were to implement something like this I think the easiest and most sensible way might be to just let the cup admin assign final points and omit levels, and then we can just display some level statistics for each level after the event is over. Ie. we could show modified playtime, which filters out runs > 2x top finish time. Also, some of the level stats work that I recently did can give some statistics for overall playtime with outliers somewhat filtered out, but it does not depend on the levels top finish time so it's generally not as accurate.

**Reply by PabloCorso** [reply]
I didn't think about having skips but it could be a nice idea depending on how many levels are selected to calculate the results. For the current skip system I believe that if for example you allow 1 skip, it will not count the worst time (or DNF) of a player to calculate his points. With 2 skips, the worst 2 times... and so on.

In this case we first want to select X levels regardless the players and times by ranking them with a specific algorithm, then we could have skips in the normal sense over those X levels.

I don't know if we are able or need a way to manually assign points, but could add a way to manually select X levels and those levels will be taken into account for the results. The points will be assigned as usual by positions (100, 85, 75, etc) with possibly some skips. This manual selection could make sense to code it into the cup system.

Knowing which levels to select by ranking them it's another story, maybe it doesn't make sense to add this specific logic into elma.online at all. But I would still will need to access all those "private" finished times to rank the levels regularly, or think another way.

If I have access to all the finishes from all levels at any moment I could get those, calculate the new levels ranking and upload the ranking somewhere else. When the cup ends, we already know the top X levels and just manually select them inside elma.online to trigger the results calculation? Is the overall idea somewhat clear?

I don't know if by playtime you mean the time played. The best ranking of the levels that I found so far is by summing the times finished, not the time played. Time played could mean that a player just stayed inside a level spying others right? In that case it is not an useful data to rank the levels.

**Comment by j-masci** [comment]
So I haven't actually looked at pretty much any of the cup code and don't know how the cups work in general. However, the easiest and most sensible way to me seems to be 1. Add an option for cup owner to exclude certain levels 2. Show a bunch of level statistics for all levels within the cup. The cup admin would then have to take care of the rest. If you have specific statistics you'd want to show, like the total playtime not including runs over 2x the best finish time, it's most likely a very easy thing to do, but it depends.

The current level stats system (which is in testing right now) also filtered out inactive runs by looking at the total run time and the # of actions (volts/turns) per minute, and throttle and brake time percentage, to identify runs that were likely AFK.

Also, estimating a levels popularity by total time played on finishes only seems pretty skewed against hard levels, ie. pipes. Also, total time played by all players might not even be good because one player can easily play a level much more than anyone else and skew the results. Also if one player does have a first place time on some level they enjoyed, why wouldn't they go and put up 50 more finishes to ensure their level gets included? Basically, any cumulative play time statistic can easily be manipulated.

Perhaps a voting system could work and could be handled outside of EOL (forums?). I think overall it needs to be based on number of players who voted (or played the level enough times, on legit runs), rather than cumulative time by all players.

I think that even in a system where total points don't totally make sense, there's still incentive to play cup levels and try to get some good finishes. A top 3 finish for example is something that not good players might like to go after on some level they enjoy even if they have no chance of getting a good overall ranking in the cup. So, there is that to think about also. 

I can't really think of the best way to include/exclude levels especially if you don't know how many levels are likely to be played in the first place and how many levels you want to end up counting towards the overall results.

**Comment by sunehs** [comment]
For the technical aspect, adding a boolean for "this event counts" that cup admin can toggle should be easy to implement. Then can make some various stats that cup admin can use to make this decision. These stats could also be interesting for normal cups, that way the effort is not wasted if the special cup isn't a success. Can be a setting for cups if those stats should be visible during the events.

For the concept itself, it could be fun if there's a lot of participation, but I feel like it could easily turn out badly as well. You never know how many levels you'll actually get, if you get just a couple levels it falls short quickly. Players may also be afraid of playing levels that'll end up not counting, so they don't play, or it'll be decided after day 1 which levels people play since everyone will try to play the most popular.

**Reply by PabloCorso** [reply]
I agree with everything :) 

The boolean sounds very nice, that's more than enough for the first point.

Having stats to access as an admin would be awesome, then I guess I could download the finished times and do the levels ranking calculation? Then I would need to share the ranked levels at Mopolauta or even a simple static site.

About the cup success, I also see the possibility of falling short. It would need some good advertising in Mopolauta, Discord and EOL at least to gather some feedback on how many people could be interested. The more players the better for sure.

On the level amount, I think it would be easier to ask designers to participate, there are currently many active and great designers around :) although any level would be accepted and that could get interesting too.

It is ambitious to say the least :) in my short experience of making cups, I can see the possibility of few people playing for real. Might need to wait for the European summer and FEM to pass before going with it...
